### PR TITLE
Partial SpikeReport code review and improvements

### PR DESCRIPTION
**Pull request by eile**

### CODE REVIEW

File: brion/plugin/spikeReportBinary.cpp [hunk 1]
```diff
@@ -159,25 +132,22 @@ bool SpikeReportBinary::handles( const SpikeReportInitData& initData )
...
 
 Spikes SpikeReportBinary::read( const float )
 {
     // In file based reports, this function reads all remaining data.
     Spikes spikes;
-    Spike* spikeArray = _memFile.getSpikes();
-    const size_t nElems = _memFile.getSpikeCount();
+    const Spike* spikeArray = _memFile->getReadableSpikes();
+    const size_t nElems = _memFile->getNSpikes();
+    _currentTime = spikeArray[nElems - 1].first +
```
Comment (kaabimg) [inline brion/plugin/spikeReportBinary.cpp]: In the case of an empty spike report file (only a header, with no spikes, this will access to an empty array

File: brion/plugin/spikeReportBinary.cpp [hunk 2]
```diff
@@ -199,24 +169,22 @@ Spikes SpikeReportBinary::readUntil( const float max )
         pushBack( spikeArray[_startIndex], spikes );
     }
 
-    if ( _startIndex == nElems )
+    if( _startIndex == nElems )
     {
-        _currentTime = UNDEFINED_TIMESTAMP;
+        _currentTime = spikeArray[nElems - 1].first +
```
Comment (kaabimg) [inline brion/plugin/spikeReportBinary.cpp]: Same comment as above : access to an empty array when _startIndex == nElems == 0

File: brion/plugin/spikeReportBinary.cpp [hunk 3]
```diff
@@ -38,113 +38,86 @@ namespace
...
+    {
+        return reinterpret_cast< Spike* >( _map.getAddress< uint8_t >() +
+                                           sizeof( Header ));
+    }
 
-bool BinaryReportFileMemMap::isMappedForRead() const
-{
-    return _mappedForRead;
-}
+    operator bool() const { return _map.getAddress() != nullptr; }
```
Comment (kaabimg) [inline brion/plugin/spikeReportBinary.cpp]: This operator is no longer used

Comment (eile) [inline brion/plugin/spikeReportBinary.cpp]: removed

File: brion/plugin/spikeReportBinary.cpp (hunk 2 above)
Comment (eile) [inline brion/plugin/spikeReportBinary.cpp]: done

File: brion/plugin/spikeReportBinary.cpp (hunk 1 above)
Comment (eile) [inline brion/plugin/spikeReportBinary.cpp]: done

**Comment by eile** [comment]
Nice tumbleweed here.

File: brion/spikeReport.h [hunk 4]
```diff
@@ -62,7 +62,7 @@ class SpikeReport : public boost::noncopyable
      * State::ended means that the report end is reached (in read mode)
      * State::failed means that a error occured when readyn or writing the report
      */
-    enum class State : uint8_t
+    enum class State : uint32_t
```
Comment (hernando) [inline brion/spikeReport.h]: What's the reason for using uint32_t? (in this case I think you can even omit it)

File: brion/plugin/spikeReportBinary.cpp [hunk 5]
```diff
@@ -236,26 +210,22 @@ void SpikeReportBinary::readSeek( const float toTimeStamp )
...
 
     if ( spikes.empty() )
         return;
 
     // create or resize the file
-    if ( !_memFile || _memFile.getSpikeCount() != totalSpikeCount )
-        _memFile.mapForReadWrite( getURI().getPath(), totalSpikeCount );
-
-    Spike* spikeArray = _memFile.getSpikes();
+    if ( _memFile->getNSpikes() != totalSpikes )
```
Comment (hernando) [inline brion/plugin/spikeReportBinary.cpp]: I preferred getSpikeCount, I don't like this sort of abbreviations.

File: brion/plugin/spikeReportBinary.cpp [hunk 6]
```diff
@@ -55,85 +53,68 @@ class Header
...
-Spike* BinaryReportFile::getSpikes() const
-{
-    return reinterpret_cast< Spike* >( _map.getAddress< uint8_t >() +
-                                       sizeof( Header ));
-}
+    Spike* writeSpikes()
+    {
+        return reinterpret_cast< Spike* >( _map.getAddress< uint8_t >() +
+                                           sizeof( Header ));
+    }
```
Comment (hernando) [inline brion/plugin/spikeReportBinary.cpp]: Rename to mutableSpikes() and constSpikes() or use getSpikes for both with const and mutable overloads.

File: brion/plugin/spikeReportBinary.cpp [hunk 7]
```diff
@@ -90,13 +90,13 @@ class BinaryReportMap
...
 
-    const Spike* readSpikes() const
+    const Spike* getReadableSpikes() const
     {
         return reinterpret_cast< const Spike* >( _map.getAddress< uint8_t >() +
                                                  sizeof( Header ));
     }
 
-    Spike* writeSpikes()
+    Spike* getWritableSpikes()
```
Comment (hernando) [inline brion/plugin/spikeReportBinary.cpp]: OK, I was looking at the commits in order to make it easier to digest.

File: brion/plugin/spikeReportBinary.cpp [hunk 8]
```diff
@@ -137,13 +140,14 @@ Spikes SpikeReportBinary::read( const float )
...
+    const Spike* spikeArray = _memFile->getReadableSpikes();
     const size_t nElems = _memFile->getNSpikes();
+    _currentTime = spikeArray[nElems - 1].first +
+                   std::numeric_limits< float >::epsilon();
 
     for ( ; _startIndex < nElems; ++_startIndex )
         pushBack( spikeArray[_startIndex], spikes );
 
-    _currentTime = UNDEFINED_TIMESTAMP;
     _state = State::ended;
```
Comment (hernando) [inline brion/plugin/spikeReportBinary.cpp]: State shouldn't be ended neither.

File: brion/plugin/spikeReportASCII.cpp [hunk 9]
```diff
@@ -48,7 +46,8 @@ Spikes SpikeReportASCII::read( const float )
     Spikes spikes;
     auto start = _lastReadPosition;
     _lastReadPosition = _spikes.end();
-    _currentTime = UNDEFINED_TIMESTAMP;
+    _currentTime = _spikes.rbegin()->first +
+                   std::numeric_limits< float >::epsilon();
```
Comment (hernando) [inline brion/plugin/spikeReportASCII.cpp]: The spec doesn't say anything about the value of getCurrentTime() when state is ended, it's undefined. For that reason, your code shouldn't rely on any specific value. My gut tells me this change is wrong if you don't constraint the specification, but referring to deltas there also seems wrong. Do you really need this change to make the incremental converter work?

Comment (hernando) [inline brion/plugin/spikeReportASCII.cpp]: The same goes for all other places in which you've done the same.

File: brion/plugin/spikeReportBinary.cpp [hunk 10]
```diff
@@ -115,6 +115,9 @@ SpikeReportBinary::SpikeReportBinary( const SpikeReportInitData& initData )
         _memFile.reset( new BinaryReportMap( getURI().getPath( )));
     else
         _memFile.reset( new BinaryReportMap( getURI().getPath(), 0 ));
+
+    if( _memFile->getNSpikes() > 0 )
+        _currentTime = _memFile->getReadableSpikes()[0].first;
```
Comment (hernando) [inline brion/plugin/spikeReportBinary.cpp]: This is the only place I find really acceptable to replace -inf by something else, but you have to update the specification of getCurrentTime accordingly

File: brion/plugin/spikeReportASCII.cpp [hunk 11]
```diff
@@ -46,8 +46,7 @@ Spikes SpikeReportASCII::read( const float )
     Spikes spikes;
     auto start = _lastReadPosition;
     _lastReadPosition = _spikes.end();
-    _currentTime = _spikes.rbegin()->first +
-                   std::numeric_limits< float >::epsilon();
+    _currentTime = UNDEFINED_TIMESTAMP;
```
Comment (hernando) [inline brion/plugin/spikeReportASCII.cpp]: Argh, this incremental reviewing approach doesn't seem to be a good idea...

I think you should make separate PRs for independent things, for example the command line argument parsing could have gone in a separate one.

File: brion/spikeReport.h (hunk 4 above)
Comment (eile) [inline brion/spikeReport.h]: uint8_t seemed like a premature optimization. Will remove type.

File: apps/CMakeLists.txt [hunk 12]
```diff
@@ -1,26 +1,27 @@
...
 set(COMPARTMENTCONVERTER_SOURCES compartmentConverter.cpp)
 set(COMPARTMENTCONVERTER_LINK_LIBRARIES Brion ${Boost_PROGRAM_OPTIONS_LIBRARY})
+
 if(TARGET BBPTestData)
   list(APPEND COMPARTMENTCONVERTER_LINK_LIBRARIES BBPTestData)
+  list(APPEND SPIKECONVERTER_LINK_LIBRARIES BBPTestData)
 endif()
 
+common_application(spikeConverter)
+common_application(morphologyConverter)
```
Comment (eile) [inline apps/CMakeLists.txt]: It allows grouping the link libs in if(testdata) above.

File: brion/plugin/spikeReportBinary.cpp (hunk 5 above)
Comment (eile) [inline brion/plugin/spikeReportBinary.cpp]: Renamed to getNumSpikes to be consistent with the rest of the API

File: brion/plugin/spikeReportBinary.cpp (hunk 6 above)
Comment (eile) [inline brion/plugin/spikeReportBinary.cpp]: are you fine with the current getWritableSpikes/getReadableSpikes?

Comment (hernando) [inline brion/plugin/spikeReportBinary.cpp]: It's fine. Using the same name for both doesn't work?
